diff --git a/node_modules/@tryghost/kg-markdown-html-renderer/lib/markdown-html-renderer.js b/node_modules/@tryghost/kg-markdown-html-renderer/lib/markdown-html-renderer.js
index ff5470f..65236ec 100644
--- a/node_modules/@tryghost/kg-markdown-html-renderer/lib/markdown-html-renderer.js
+++ b/node_modules/@tryghost/kg-markdown-html-renderer/lib/markdown-html-renderer.js
@@ -87,9 +87,9 @@ const selectRenderer = function (options) {
         return markdownIt;
     }
 };
-
+const getRandom = () => Math.random().toString(36).slice(2, 9);
 module.exports = {
     render: function (markdown, options = {}) {
-        return selectRenderer(options).render(markdown);
+        return selectRenderer(options).render(markdown, {docId: getRandom()});
     }
 };
