diff --git a/cjs/color-utils.js b/cjs/color-utils.js
index 38e1145b2dfc18f35ccb5b551d0c644bc90a8968..7cc46b5e793654821f805d283a39844c04a525fd 100644
--- a/cjs/color-utils.js
+++ b/cjs/color-utils.js
@@ -37,4 +37,4 @@ function textColorForBackgroundColor(background) {
     // shared with Portal https://github.com/TryGhost/Portal/blob/317876f20d22431df15e655ea6cc197fe636615e/src/utils/contrast-color.js#L26-L29
     const yiq = (backgroundColor.red() * 0.299 +
         backgroundColor.green() * 0.587 +
-        backgroundColor.b() * 0.114);
+        backgroundColor.blue() * 0.114);
diff --git a/es/color-utils.js b/es/color-utils.js
index 0de18a4542e36e87a11c5a47537664cd286dfc96..a8f38c418fd57e1c1834c41836e2f48274dcf58e 100644
--- a/es/color-utils.js
+++ b/es/color-utils.js
@@ -2160,4 +2160,4 @@ function textColorForBackgroundColor(background) {
     // shared with Portal https://github.com/TryGhost/Portal/blob/317876f20d22431df15e655ea6cc197fe636615e/src/utils/contrast-color.js#L26-L29
     const yiq = (backgroundColor.red() * 0.299 +
         backgroundColor.green() * 0.587 +
-        backgroundColor.b() * 0.114);
+        backgroundColor.blue() * 0.114);
diff --git a/src/color-utils.ts b/src/color-utils.ts
index d797083148b521080aa926e4a1b93577890954b6..d8dcb1de884b76316bb7b43be02ff1bc48663d35 100644
--- a/src/color-utils.ts
+++ b/src/color-utils.ts
@@ -50,4 +50,4 @@ export function textColorForBackgroundColor(background: string | Color): Color {
     const yiq = (
         backgroundColor.red() * 0.299 +
         backgroundColor.green() * 0.587 +
-        backgroundColor.b() * 0.114
+        backgroundColor.blue() * 0.114
